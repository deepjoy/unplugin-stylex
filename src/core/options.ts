import type { StylexOptions, UnpluginStylexOptions } from '@/types'
import { isDevelopment } from './constants'

export function getOptions(options: UnpluginStylexOptions): Required<UnpluginStylexOptions> {
  const stylex = options.stylex || ({} as StylexOptions)
  const isDev = options.dev || isDevelopment
  if(options.invalidExts && options.validExts) {
    throw new Error('Only one of invalidExts or validExts may be specified at a time')
  }

  return {
    ...options,
    dev: isDev,
    invalidExts: options.invalidExts ? ['.json', '.html', '.jade', '.json5', ...(options.invalidExts ?? [])] : [],
    // .js, .jsx, .mjs, .cjs, .ts, .tsx, .mts, .cts
    validExts: options.validExts ?? /\.[mc]?[jt]sx?$/,
    stylex: {
      filename: stylex.filename || 'stylex.css',
      stylexImports: stylex.stylexImports || ['@stylexjs/stylex'],
      runtimeInjection: stylex.runtimeInjection ?? isDev,
      aliases: stylex.aliases,
      useCSSLayers: stylex.useCSSLayers || false,
      unstable_moduleResolution: stylex.unstable_moduleResolution || { type: 'commonJS', rootDir: process.cwd() },
      babelConfig: {
        babelrc: (stylex.babelConfig || {}).babelrc || false,
        plugins: (stylex.babelConfig || {}).plugins || [],
        presets: (stylex.babelConfig || {}).presets || [],
      },
      ...stylex,
    },
  }
}
