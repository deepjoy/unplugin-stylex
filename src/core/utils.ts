import type { UnpluginStylexOptions, BabelConfig } from '@/types'
import { isDevelopment } from './constants'

export function getOptions(options: UnpluginStylexOptions = ({} as UnpluginStylexOptions)) {
  const stylex = options.stylex || {}

  return {
    ...options,
    dev: options.dev || isDevelopment,
    include: options.include || [/\.[jt]sx?$/],
    exclude: options.exclude || [/node_modules/, /\.git/],
    enforce: options.enforce || 'pre',
    stylex: {
      ...stylex,
      filename: stylex.filename || 'stylex.css',
      stylexImports: stylex.stylexImports || ['stylex', '@stylexjs/stylex'],
      useCSSLayers: stylex.useCSSLayers || false,
      unstable_moduleResolution: stylex.unstable_moduleResolution || { type: 'commonJS', rootDir: process.cwd() },
      babelConfig: {
        babelrc: (stylex.babelConfig || {}).babelrc || false,
        plugins: (stylex.babelConfig || {}).plugins || [],
        presets: (stylex.babelConfig || {}).presets || [],
      },
    },
  }
}
