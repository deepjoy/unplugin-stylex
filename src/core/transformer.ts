import * as fs from 'fs/promises'
import * as path from 'path'

import { transformAsync } from '@babel/core'
import flowSyntaxPlugin from '@babel/plugin-syntax-flow'
import jsxSyntaxPlugin from '@babel/plugin-syntax-jsx'
import typescriptSyntaxPlugin from '@babel/plugin-syntax-typescript'
import stylexBabelPlugin from '@stylexjs/babel-plugin'

export async function transformer(context) {
  const { id, source, options } = context
  const stylex = options.stylex
  const extname = path.extname(id)
  const stylexRules = {}

  if (stylex.stylexImports.some((importName) => source.includes(importName))) {
    const originSource = stylex.babelConfig.babelrc
      ? await fs.readFile(id, 'utf-8')
      : source

    const { code, map, metadata } = await transformAsync(
      originSource,
      {
        babelrc: stylex.babelConfig.babelrc,
        filename: id,
        plugins: [
          ...(stylex.babelConfig.plugins || []),
          extname === '.ts'
            ? typescriptSyntaxPlugin
            : extname === '.tsx'
              ? [typescriptSyntaxPlugin, { isTSX: true }]
              : flowSyntaxPlugin,

          jsxSyntaxPlugin,
          [
            stylexBabelPlugin,
            {
              dev: options.dev,
              unstable_moduleResolution: stylex.unstable_moduleResolution,
              importSources: stylex.stylexImports,
              ...stylex,
            }
          ],
        ],
        presets: stylex.babelConfig.presets,
      }
    )

    if (metadata.stylex && metadata.stylex.length > 0) {
      stylexRules[id] = metadata.stylex
    }

    if (!stylex.babelConfig.babelrc) {
      return { code, map, stylexRules }
    }

    return { code, stylexRules }
  }

  return { code: source }
}
