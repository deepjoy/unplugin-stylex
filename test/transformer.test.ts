import { describe, expect, it } from 'vitest'
import { transformer } from '../src/core/transformer'

[ 'javascript', 'typescript' ].forEach((name) => {
  describe(name, () => {
    it('basic', async () => {
      const context = {
        id: '',
        source: '',
        options: {
          dev: true,
          stylex: {},
        },
      }

      const { code } = await transformer(context)
      expect(code).toMatchSnapshot()
    })

    it('useCSSLayer', async () => {
      const stylexOptions = {
        useCSSLayer: true,
      }

      const context = {
        id: '',
        source: '',
        options: {
          dev: true,
          stylex: stylexOptions,
        },
      }

      const { code } = await transformer(context)
      expect(code).toMatchSnapshot()
    })
  })
})
