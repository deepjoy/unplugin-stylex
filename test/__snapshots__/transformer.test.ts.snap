// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`javascript > basic 1`] = `""`;

exports[`javascript > useCSSLayer 1`] = `""`;

exports[`typescript > basic 1`] = `""`;

exports[`typescript > useCSSLayer 1`] = `""`;
